src/ORM/Paginator.php

```diff
@@ -0,0 +1,404 @@
+     *   that key's settings will be used for pagination instead of the general ones.
+     * @param array $settings The settings to merge with the request data.
+     * @return array Array of merged options.
+     */
+    public function mergeOptions($alias, $settings)
+    {
+        $defaults = $this->getDefaults($alias, $settings);
+        $otherParams = array_intersect_key($this->_params, array_flip($this->_config['whitelist']));
+        return array_merge($defaults, $otherParams);
+    }
+
+    /**
+     * Sets pagination params that get merged in the next paginate() call.
+     *
+     * @param string $key Key to set.
+     * @param mixed $value Value to set.
+     * @throws \InvalidArgumentException
+     * @return void
+     */
+    public function setParams($name, $value = null) {
+        if (!is_array($name) && !empty($value)) {
+            $name = [$name => $value];
+        }
+        foreach ($name as $key => $value) {
+            $this->_params[$key] = $value;
+        }
+    }
+
+    /**
+     * Get the default settings for a $model. If there are no settings for a specific model, the general settings
+     * will be used.
+     *
+     * @param string $alias Model name to get default settings for.
+     * @param array $defaults The defaults to use for combining settings.
+     * @return array An array of pagination defaults for a model, or the general settings.
+     */
+    public function getDefaults($alias, $defaults)
+    {
+        if (isset($defaults[$alias])) {
+            $defaults = $defaults[$alias];
+        }
+        if (isset($defaults['limit']) &&
+            (empty($defaults['maxLimit']) || $defaults['limit'] > $defaults['maxLimit'])
+        ) {
+            $defaults['maxLimit'] = $defaults['limit'];
+        }
+        return $defaults + $this->config();
+    }
+
+    /**
+     * Validate that the desired sorting can be performed on the $object. Only fields or
+     * virtualFields can be sorted on. The direction param will also be sanitized. Lastly
+     * sort + direction keys will be converted into the model friendly order key.
+     *
+     * You can use the whitelist parameter to control which columns/fields are available for sorting.
+     * This helps prevent users from ordering large result sets on un-indexed values.
+     *
+     * If you need to sort on associated columns or synthetic properties you will need to use a whitelist.
+     *
+     * Any columns listed in the sort whitelist will be implicitly trusted. You can use this to sort
+     * on synthetic columns, or columns added in custom find operations that may not exist in the schema.
+     *
+     * @param Table $object The model being paginated.
+     * @param array $options The pagination options being used for this request.
+     * @return array An array of options with sort + direction removed and replaced with order if possible.
+     */
+    public function validateSort(Table $object, array $options)
```

Comment:
I've tried that and got this:

```
1) Cake\Test\TestCase\Controller\Component\PaginatorComponentTest::testPaginateQuery
Argument 1 passed to Cake\ORM\Paginator::paginate() must implement interface Cake\Datasource\RepositoryInterface, instance of Mock_Query_143691e0 given,
```

What's the recommended way to resolve this?